src/Components/Components/ref/Microsoft.AspNetCore.Components.netcoreapp.cs

```diff
@@ -242,6 +242,10 @@ public partial interface IHandleEvent
     {
         System.Threading.Tasks.Task HandleEventAsync(Microsoft.AspNetCore.Components.EventCallbackWorkItem item, object? arg);
     }
+    public partial interface ILazyLoader
+    {
+        System.Threading.Tasks.Task<System.Collections.Generic.IEnumerable<System.Reflection.Assembly>> LoadAssembliesAsync(System.Collections.Generic.IEnumerable<string> assembliesToLoad);
+    }
```

Comment:
I'm not sure if it makes sense for this to live here. I imagine we need to provide an implementation that is compatible with prerendering, but this doesn't seem right.

@SteveSandersonMS do you have any thoughts? 

We keep piling up these types of things in MAC, so it might be time we find a different way to do initialization maybe?

Reply:
One option is that developers doing prerendering+WASM could do the following, which I know @javiercn will hate:

```cs
@inject IServiceProvider serviceProvider

@code {
void OnInitializing()
{
    if (Platform.IsWebAssembly)
    {
        loader = serviceProvider.GetService<IWebAssemblyLazyLoader>();
    }
}

async Task HandleNavigate(args)
{
    if (Platform.IsWebAssembly)
    {
        var assembliesToLoad = CalculateFrom(args.Url);
        await loader.LoadAssemblies(assembliesToLoad);
    }
}
}
```

... then it will just work with prerendering. Might be a reasonable first step which we could potentially improve on in the future if customers demand it. It's only an issue for people doing prerendering - everyone else can ignore this.

I know it would be better if we had a conditional `@inject` syntax that didn't throw if no matching service was registered, then there would be no need to inject the service provider.

Alternatively we could have a `NullAssemblyLoader` that gets registered in `M.A.C.Server` which throws if you try to call its methods, though that does require defining an abstraction in `M.A.Components`.

Reply:
I died a bit on the inside.

I think it's not that complicated actually. We can actually make it work in a similar way as https://github.com/dotnet/aspnetcore/blob/c1866c2b61361e601a448a82607d8879ad01696a/src/JSInterop/Microsoft.JSInterop/src/Infrastructure/DotNetDispatcher.cs#L380-L396

We just need the WASM implementation to look for assemblies that are already loaded in the app domain, since on a server context they'll be referenced like regular assemblies and copied out to the output folder.

Then the only missing piece is an extension method over the service collection to register it, or to make the type public so that users can register it themselves.

Reply:
We can ditch this interface, have a public abstract type instead and require the user to register it on DI if they are using wasm + prerendering.

Reply:
> We can ditch this interface, have a public abstract type instead and require the user to register it on DI if they are using wasm + prerendering.

+1 great plan

Reply:
> have a public abstract type instead and require the user to register it on DI if they are using wasm + prerendering

Also the implementation should throw if you ask it to load any assemblies while not running on WebAssembly. This will force developers to put a platform `if` check around calls to it, which avoids confusion. We don't want people thinking it makes sense to do actual lazy loading on the server.

This seems to solve all the issues:

 * No need for a new abstraction in `M.A.Components`
 * No need to inject service provider into the component hosting the router
 * No confusion thinking it makes sense to lazy-load on the server

Reply:
> Also the implementation should throw if you ask it to load any assemblies while not running on WebAssembly. This will force developers to put a platform `if` check around calls to it, which avoids confusion. We don't want people thinking it makes sense to do actual lazy loading on the server.

I really dislike forcing the users to put an If else in their code for something we can provide a reasonable server-side implementation. As a compromise, what if we add a public constructor that allows you to pass in a list of pre populated assemblies for server scenarios and we only throw if the assembly is not contained in that list?

Reply:
Reading this thread linearly and posting my thoughts as I go.

I actually did try the approach if injecting the service provider and doing a check at runtime to see if you could load assemblies. I decided against it because it seems to clunky and I don't like the fact that injecting the service provider instead of the interface makes it harder to understand what the code is doing at a glance.

> I know it would be better if we had a conditional @inject syntax that didn't throw if no matching service was registered, then there would be no need to inject the service provider.

Yes, please! If we're keen on adding this feature (regardless of its affect on this work), I can open an issue for this.

> We can ditch this interface, have a public abstract type instead and require the user to register it on DI if they are using wasm + prerendering.

I'm fine with this. I hadn't considered the option of "make the user register it in the DI" while working on this but this makes sense because lazy-loading is an optional, not built-in feature.

> This will force developers to put a platform if check around calls to it, which avoids confusion. We don't want people thinking it makes sense to do actual lazy loading on the server.

I'm a little torn about making the user right conditional checks in their code. Would it suffice to just throw a warning under development and have the server-side implementation no-op?

Reply:
> I'm a little torn about making the user right conditional checks in their code. Would it suffice to just throw a warning under development and have the server-side implementation no-op?

The server side implementation can't no-op, you won't be able to prerender lazy-loaded routes in that case.

Reply:
> > I know it would be better if we had a conditional @Inject syntax that didn't throw if no matching service was registered, then there would be no need to inject the service provider.
> 
> Yes, please! If we're keen on adding this feature (regardless of its affect on this work), I can open an issue for this.

I'd rather we don't add a new concept for this if we can avoid it.

I'd rather us provide users with abstractions that can work on the contexts that we care about and encourage library authors to do the same over forcing users make conditional checks all over their codebases.

Reply:
> forcing users make conditional checks all over their codebases

Sidenote: platform checks will be a way of life in .NET 5. The new analyzer that Immo is working on will provide compile-time feedback about which BCL APIs are usable on WebAssembly and which aren't. It does code flow analysis so you only get warnings about code that isn't guarded by suitable "if platform" checks.

> The server side implementation can't no-op, you won't be able to prerender lazy-loaded routes in that case.

Not sure what that means. On the server, all the assemblies are loaded all the time right from when the app starts up.

Were you thinking that "lazy loading" means that some of the assemblies aren't referenced from the app's main assembly? Because that's not the design we've settled on. We settled on regular package/project references at compile time, as it's the only thing that makes linking work correctly.

Reply:
To clarify, by "no-op" I mean it won't actually call `Assembly.Load` or any similar variant.

It might still return the assemblies that will be loaded or through an exception but it won't actually load anything because they've already been loading.

> Were you thinking that "lazy loading" means that some of the assemblies aren't referenced from the app's main assembly?

Nope, clarified above. We anticipate that all assemblies will be referenced from the app to support how we implemented build-time support for this.

Reply:
> To clarify, by "no-op" I mean it won't actually call `Assembly.Load` or any similar variant.

Yes, the assemblies should be already loaded, so the only "work" here is to provide the proper answer based on the loaded assemblies or throw if it is not possible to do so.

We should not call Assembly.Load in any of its variants, hence why I pointed out the code we use in DotNetDispatcher